### PR TITLE
Engineering Diffraction GUI - run GSASIIRefineFitPeaks in a separate thread

### LINKED ISSUE #21843: Engineering GUI GSAS Refinement bug fixes
There are a few remaining bugs to fix up in the GSAS GUI to fix up before release.
So far we are aware of:
* The GUI freezes when you run the algorithm (this can be solved by running alg in a separate thread - Not the GUI thread)

### PR DESCRIPTION
**Pull request by josephframsay**
Previously, running GSASIIRefineFitPeaks synchronously from the GSAS tab of the ED GUI locked the GUI. This PR takes care of moving the algorithm to run in a separate thread, and hence stop locking the GUI.

**To test:**

You'll need GSAS-II installed. Install it using one of the scripts in `scripts/GSAS-II` or follow the instructions [here](https://subversion.xray.aps.anl.gov/trac/pyGSAS). Test data is in [GSASIIRefineFitPeaks.zip](https://github.com/mantidproject/mantid/files/1739753/GSASIIRefineFitPeaks.zip)

Start the ED GUI by going to `Interfaces -> Diffraction -> Engineering Diffraction`. You'll be prompted to enter an RB number if you haven't opened the GUI before - do so then go to the `GSAS-II Refinement` tab.

Load the run by clicking the first **Browse** button, selecting `ENGINX_280625_focused_bank_1.nxs` from the zip, and clicking **Load**. `280625_1` should appear in the list in the bottom right of the GUI - click this and the run should appear in the plot.

Select `template_ENGINX_241391_236516_North_bank.prm` for the instrument parameter file, all three `.cif` files for the phase files, a new location to save the gsas project (`.gpx`) to, and your GSAS-II installation directory (containing `GSASIIscriptable.py`), all with their respective browse buttons. Feel free to fiddle with other controls if you feel so inclined.

Click **Run Refinement**. This time the GUI should not lock up, instead all the inputs should be disabled, and the algorithm should run in a separate thread. Make sure they are re-enabled once the algorithm has finished. Try cancelling the algorithm, closing the GUI before the algorithm has finished, and anything else you can think of that might break things.

Also code review (please!)

Fixes #21843 

**Release Notes** 
[Here](https://github.com/mantidproject/mantid/pull/21916/commits/0e513059bd39cf01bbbea4f650a66d027d0f1dc7)

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [x] Are the unit tests small and test the class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] Do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by ElliotAOram** [comment]
@Mantid-Matthew you can follow the test instructions on this [Pull Request](https://github.com/mantidproject/mantid/pull/21804). I believe this is left in a state where the problem with the GUI freezing has been resolved, just the tests and error handling need a fix up. Let me know if you have any questions and I can try to answer them!

**Comment by NickDraper** [comment]
@ElliotAOram @Mantid-Matthew  has got an urgent issue in from one of the SANS instruments, so I've removed him from this PR.  We will need to find someone else.

**Comment by NickDraper** [comment]
Moving out of this release as more work is needed here by the original author than we have time for in this release.

**Comment by josephframsay** [comment]
Accidentally closed with deletion of `release-v3.12` branch

**Comment by josephframsay** [comment]
@NickDraper Would this be considered too high-risk for the patch release?

File: qt/scientific_interfaces/EnggDiffraction/EnggDiffGSASFittingModel.cpp [hunk 1]
```diff
@@ -19,20 +19,37 @@ std::string stripWSNameFromFilename(const std::string &fullyQualifiedFilename) {
...
 
+std::string refinementMethodToString(
+    const MantidQt::CustomInterfaces::GSASRefinementMethod &method) {
+  switch (method) {
+  case MantidQt::CustomInterfaces::GSASRefinementMethod::PAWLEY:
+    return "Pawley refinement";
+  case MantidQt::CustomInterfaces::GSASRefinementMethod::RIETVELD:
+    return "Rietveld refinement";
+  default:
+    return "Unknown refinement method: contact development team";
```
Comment (soininen) [inline qt/scientific_interfaces/EnggDiffraction/EnggDiffGSASFittingModel.cpp]: Would it make sense to throw here?

File: qt/scientific_interfaces/EnggDiffraction/IEnggDiffGSASFittingObserver.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+#define MANTIDQTCUSTOMINTERFACES_ENGGDIFFRACTION_GSASFITTINGOBSERVER_H_
+
+#include "GSASIIRefineFitPeaksOutputProperties.h"
+
+namespace MantidQt {
+namespace CustomInterfaces {
+
+class IEnggDiffGSASFittingObserver {
+
+public:
```
Comment (soininen) [inline qt/scientific_interfaces/EnggDiffraction/IEnggDiffGSASFittingObserver.h]: This class has `virtual` methods. How about a `virtual` destructor as well? Could be declared `default`.

File: qt/scientific_interfaces/EnggDiffraction/IEnggDiffGSASFittingPresenter.h [hunk 3]
```diff
@@ -1,10 +1,12 @@
...
+#include "IEnggDiffGSASFittingObserver.h"
+
 namespace MantidQt {
 namespace CustomInterfaces {
 
-class IEnggDiffGSASFittingPresenter {
+class IEnggDiffGSASFittingPresenter : public IEnggDiffGSASFittingObserver {
 
 public:
   virtual ~IEnggDiffGSASFittingPresenter() = default;
```
Comment (soininen) [inline qt/scientific_interfaces/EnggDiffraction/IEnggDiffGSASFittingPresenter.h]: The destructor could be declared `virtual` in the new parent class and removed here.

**Review by soininen — CHANGES_REQUESTED** [review]
I had some problems saying that GSASII could not be imported but these got solved after I set up correct proxies for subversion. Nothing that could be solved in Mantid, I suppose. Anyway, the refinement seems to run nicely in the background and nothing crashes in unrecoverable way if I delete workspaces or close the interface in the middle of the refinement process. There are a few concerns regarding the code but nothing major, see below.

**Review by soininen — APPROVED** [review]

**Pull request merged**